src/main/java/org/checkerframework/specimin/Slicer.java

```diff
@@ -0,0 +1,212 @@
+   * same value, and {@code ==} yields the same reference.
+   *
+   * <p>An {@code IdentityHashMap} is safe to use because we use the same compilation units
+   * everywhere, so all corresponding nodes will have the same reference.
+   */
+  private final Set<Node> slice = Collections.newSetFromMap(new IdentityHashMap<>());
+
+  /** The slice of generated symbols. */
+  private final Set<UnsolvedSymbolAlternates<?>> generatedSymbolSlice = new HashSet<>();
+
+  /** The Nodes that need to be processed in the main algorithm. */
+  private final Deque<Node> worklist = new ArrayDeque<>();
+
+  private final Set<Node> postProcessingWorklist = new HashSet<>();
+
+  private final Map<String, Path> existingClassesToFilePath;
+  private final String rootDirectory;
+  private final Map<String, CompilationUnit> toSlice;
+
+  private final UnsolvedSymbolGenerator unsolvedSymbolGenerator;
+  private final TypeRuleDependencyMap typeRuleDependencyMap;
+
+  public Slicer(
+      Map<String, Path> existingClassesToFilePath,
+      String rootDirectory,
+      Map<String, CompilationUnit> toSlice,
+      TypeRuleDependencyMap typeRuleDependencyMap,
+      UnsolvedSymbolGenerator unsolvedSymbolGenerator) {
+    this.existingClassesToFilePath = existingClassesToFilePath;
+    this.rootDirectory = rootDirectory;
+    this.toSlice = toSlice;
+    this.typeRuleDependencyMap = typeRuleDependencyMap;
+    this.unsolvedSymbolGenerator = unsolvedSymbolGenerator;
+  }
+
+  public void slice() throws IOException {
+    Set<CompilationUnit> usedCompilationUnits = new HashSet<>();
+
+    // Step 1: build the slice; see which nodes to keep
+    while (!worklist.isEmpty()) {
+      Node element = worklist.remove();
+      handleElement(element, usedCompilationUnits);
+    }
+
+    if (!generatedSymbolSlice.isEmpty()) {
+      // Step 2: Handle the post-processing worklist
+      Iterator<Node> ppwIterator = postProcessingWorklist.iterator();
+      while (ppwIterator.hasNext()) {
+        Node element = ppwIterator.next();
+        AddInformationResult result = unsolvedSymbolGenerator.addInformation(element);
+        generatedSymbolSlice.addAll(result.toAdd());
+        generatedSymbolSlice.removeAll(result.toRemove());
+      }
+    }
+
+    // Step 3: add all used compilation units
+    for (CompilationUnit used : usedCompilationUnits) {
+      toSlice.put(
+          qualifiedNameToFilePath(used.getPrimaryType().get().getFullyQualifiedName().get())
+              .toString(),
+          used);
+    }
+
+    // Step 4: remove all nodes except the ones marked for preservation OR package/import
+    // declarations
+    for (CompilationUnit cu : toSlice.values()) {
+      sliceCompilationUnit(cu);
+    }
+  }
+
+  public Set<UnsolvedSymbolAlternates<?>> getGeneratedSymbolSlice() {
+    return generatedSymbolSlice;
+  }
+
+  public void addToWorklist(Node node) {
+    worklist.add(node);
+  }
```

Comment:
Why do these methods (this one and `getGeneratedSymbolSlice()`) need to be public? They seem like class internals.